### PR TITLE
Woo Shipping Labels: purchase labels

### PR DESCRIPTION
**Pull request by hichamboushaba**
This adds support for purchasing shipping labels.

To test:
1. Create an order that satisfies shipping label creation in your store.
2. Open the example app.
3. Click on Woo -> Shipping Labels.
4. Click on purchase a shipping label.
5. Enter the order and package details.
6. Confirm that the shipping label details are printed in the console.

### CODE REVIEW

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/model/shippinglabels/WCShippingLabelMapper.kt [hunk 1]
```diff
@@ -32,4 +36,36 @@ class WCShippingLabelMapper
...
+                currency = labelItem.currency ?: ""
+                productNames = labelItem.productNames.toString()
+                productIds = labelItem.productIds.toString()
+                refund = labelItem.refund.toString()
+                dateCreated = labelItem.dateCreated?.toString() ?: ""
+
+                localOrderId = orderId
+                formData = gson.toJson(FormData(origin = origin, destination = destination))
+
+                localSiteId = site.id
```
Comment (hichamboushaba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/model/shippinglabels/WCShippingLabelMapper.kt:67]: Some details are missing from the actual API response, there is a task to rework the API to include them.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/model/shippinglabels/WCShippingLabelMapper.kt [hunk 2]
```diff
@@ -32,4 +36,36 @@ class WCShippingLabelMapper
...
+                packageName = labelItem.packageName ?: ""
+                rate = labelItem.rate?.toFloat() ?: labelItem.refundableAmount?.toFloat() ?: 0F
+                refundableAmount = labelItem.refundableAmount?.toFloat() ?: 0F
+                currency = labelItem.currency ?: ""
+                productNames = labelItem.productNames.toString()
+                productIds = labelItem.productIds.toString()
+                refund = labelItem.refund.toString()
+                dateCreated = labelItem.dateCreated?.toString() ?: ""
+
+                localOrderId = orderId
```
Comment (hichamboushaba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/model/shippinglabels/WCShippingLabelMapper.kt:64]: the `localOrderId` is actually referencing the remote order ID, so what we are doing here is correct, we should rename it in a separate PR.
I've filed a separate issue for this #1908

File: example/src/main/java/org/wordpress/android/fluxc/example/ui/shippinglabels/WooShippingLabelFragment.kt [hunk 3]
```diff
@@ -221,89 +225,68 @@ class WooShippingLabelFragment : Fragment() {
         get_shipping_rates.setOnClickListener {
             selectedSite?.let { site ->
                 coroutineScope.launch {
-                    prependToLog("Loading shipping data...")
+                    val orderId = showSingleLineDialog(requireActivity(), "Enter order id:", isNumeric = true)
+                            ?.toLong() ?: return@launch
```
Comment (hichamboushaba) [inline example/src/main/java/org/wordpress/android/fluxc/example/ui/shippinglabels/WooShippingLabelFragment.kt:229]: I added this one, as some orders may not satisfy the shipping labels creation conditions, so it's better to use a specific order for the tests.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/model/shippinglabels/WCShippingLabelPackageData.kt [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package org.wordpress.android.fluxc.model.shippinglabels
+
+import com.google.gson.annotations.SerializedName
+
+data class WCShippingLabelPackageData(
```
Comment (hichamboushaba) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/model/shippinglabels/WCShippingLabelPackageData.kt:5]: I didn't find a better name, if you have one, please let me know 😉

**Review by 0nko — COMMENTED** [review]
When I was testing the purchase I got the following exception after entering the package dimensions:

```
java.util.NoSuchElementException: List is empty.
        at kotlin.collections.CollectionsKt___CollectionsKt.first(_Collections.kt:212)
        at org.wordpress.android.fluxc.example.ui.shippinglabels.WooShippingLabelFragment$onViewCreated$8$$special$$inlined$let$lambda$1.invokeSuspend(WooShippingLabelFragment.kt:360)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```
         
It looks like there are no rates available for the given addresses, which is strange, since I use the same order for testing without any changes and it does give me rates in the Woo app.

**Comment by hichamboushaba** [comment]
Maybe it doesn't have any rates for the package "medium_flat_box_top" that we hardcode, did you test just fetching the rates and see what it gives?

I'll update the logic to log an error if it's empty instead of crashing.

**Comment by hichamboushaba** [comment]
@0nko I added a check to see if the list of rates is not empty before trying to access it in 9f8dfd3, I'm also caching the purchased labels after purchase in 7da18c5

Ready for another round

**Review by 0nko — APPROVED** [review]
Nice work, @hichamboushaba! :shipit:

**Pull request merged**